src/asynchronous/shutdown.rs

```diff
@@ -144,24 +144,25 @@ impl Notifier {
     /// Wait for all [`Waiter`]s to drop.
     pub async fn wait_all_exit(&self) -> Result<(), Elapsed> {
         //debug_assert!(self.shared.is_shutdown());
-        if self.waiters() == 0 {
-            return Ok(());
-        }
-        let wait = self.wait();
-        if self.waiters() == 0 {
-            return Ok(());
-        }
-        wait.await
-    }
-
-    async fn wait(&self) -> Result<(), Elapsed> {
         if let Some(tm) = self.wait_time {
-            timeout(tm, self.shared.notify_exit.notified()).await
+            timeout(tm, self.wait()).await
         } else {
-            self.shared.notify_exit.notified().await;
+            self.wait().await;
             Ok(())
         }
     }
+
+    async fn wait(&self) {
```

Comment:
```suggestion
    pub async fn wait_all_exit(&self) -> Result<(), Elapsed> {
        while self.waiters() > 0 {
            let wait = self.wait(self.shared.notify_exit.notified());
            ....
        }
    }
    async fn wait<F>(&self, future: F) -> Result<F::Output, Elapsed>
        where F: Future
    {
        if let Some(tm) = self.wait_time {
            timeout(tm, future).await
        } else {
            Ok(future.await)
        }
    }
```
`wait()` is just waiting, do `while self.waiters()` in `wait_all` ?

Reply:
This doesn't fix the issue, you need to check the waiters count after the notified() object is created but before calling await on it. Let's take an example with two threads running in parallel:

```
Thread 1  (executing last waiter.drop())        |               Thread 2 (in wait_all_exit() call)
                                                |
self.shared.waiters = 0                         |
                                                |
self.shared.notify_exit.notify_waiters()        |               
                                                |
                                                |               future = self.shared.notify_exit.notified();
                                                |
                                                |               future.await;
```

The notify_waiters() call doesn't store any permit, it only unblocks existing notified() objects: https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html#method.notify_waiters . 

If the notified object is created after that call, await will block forever. There won't be anyone left to unblock it 
because notify_waiters() is called only when the last waiter is dropped (unless a subscribe() call and a drop() happens in the meantime). 

The sync.Notify struct uses internally seq_cst operations on an AtomicUsize state value. This means that they form a happens before relationship between notified() and notify_waiters() calls, anything that happened before the call to notify_waiters() in thread 1, will be visible for the thread 2 after the notified() call. 

Hence, we can test for waiters == 0 just after the notified() call:

1. If the value is 0 awaiting at this point would be dangerous because of the scenario described above, but we can just exit. 
2. If the value is not 0 then that means there is still a reference somewhere and we can expect a notify_waiters() call on another thread later on, which will unblock our existing notified() object so we can safely await.

By adding a check for the waiters count right before await happens we can protect against this hang.

Reply:
Thanks for the explanation, very detailed. I know how the race happens and have read the documentation.

> The Notified future is guaranteed to receive wakeups from notify_waiters() as soon as it has been created, even if it has not yet been polled.

The problem is that `Notified` need to be created first, then it makes sense to test waiters. So we can move `notified()` forward.

My point is that the `wait` function is a helper function, and not public api. It maybe used to unit test or other intenal situation. The `wait_all_exit` is provided to users to ensure concurrency safety.

Reply:
There are two reasons for moving the bulk of the code in wait():

1. Let's inline wait in the snippet you suggested:

```
 pub async fn wait_all_exit(&self) -> Result<(), Elapsed> {
        while self.waiters() > 0 {
            let future = self.shared.notify_exit.notified();
            if let Some(tm) = self.wait_time {
                let wait = timeout(tm, future).await
            } else {
                let wait = Ok(future.await)
            }
            ....
        }
    }
```

Between the notified() call and the await there is no check for waiters count so awaiting can still block forever if waiters count is 0. The solution is to add the test for waiters() at the beginning of wait() function. This solution while it might seem ok it doesn't work because of the second issue.

2. The timeout, when specified, should apply to the outer while loop as well. Otherwise the code might loop multiple times, beyond the specified timeout, which is incorrect.

Reply:
The `......` is the code that leave it as is for yours.
```suggestion
        pub async fn wait_all_exit(&self) -> Result<(), Elapsed> {
	        while self.waiters() > 0 {
	            let wait = self.wait(self.shared.notify_exit.notified());
	            if self.waiters() == 0 {
                        return;
                     }
                     wait.await;
	      }
        }
```

Reply:
This won't work, it doesn't resolve the second issue that is related the timeout. The timeout should be applied to the whole while loop because it can iterate multiple times, in case someone calls subscribe() and the waiters() goes from 0 to >0. The idea is to move the while block inside the future that is passed to timeout() function. By moving it there, you'll get to the same piece of code I initially wrote.

Reply:
Thanks for your comment.
I'm going to take some time to review this.